quickwit-indexing/src/source/logrotate_source.rs

```diff
@@ -0,0 +1,984 @@
+// Copyright (C) 2022 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::BTreeMap;
+use std::io;
+use std::io::SeekFrom;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use std::time::Duration;
+
+use anyhow::Context;
+use async_trait::async_trait;
+use glob::Pattern as GlobPattern;
+use itertools::Either;
+use quickwit_actors::{ActorExitStatus, Mailbox};
+use quickwit_config::LogRotateSourceParams;
+use quickwit_metastore::checkpoint::{CheckpointDelta, PartitionId, Position, SourceCheckpoint};
+use serde::Serialize;
+use tokio::fs::{self, File};
+use tokio::io::{AsyncBufReadExt, AsyncRead, AsyncSeekExt, BufReader};
+use tracing::{debug, trace, warn};
+
+use crate::actors::Indexer;
+use crate::models::RawDocBatch;
+use crate::source::{Source, SourceContext, TypedSourceFactory};
+
+/// Cut a new batch as soon as we have read BATCH_NUM_BYTES_THRESHOLD.
+const BATCH_NUM_BYTES_THRESHOLD: u64 = 500_000u64;
+
+/// Default name when no log file has been rotated yet. It must be before any name
+/// any log file may have. This works under the assumption only printable name will
+/// be used. Alternatively we could use \0 but it may mess with some debug print.
+const START_FILE: &str = "        ";
+
+/// Time to sleep when there is nothing to do.
```

Comment:
Can we make "nothing to do" more explicit?
`when we reach the end of the current file and ...`